internal/controller/githubissue_controller.go

```diff
@@ -0,0 +1,143 @@
+	if err := r.Get(ctx, req.NamespacedName, issueObject); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			log.Error("unable to fetch issue object", zap.Error(err))
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	splitUrl := strings.Split(issueObject.Spec.Repo, "/")
+	owner := splitUrl[3]
+	repo := splitUrl[4]
+
+	log.Info(fmt.Sprintf("attempting to get issues from %s/%s", owner, repo))
+	gitHubIssue, err := r.FindIssue(ctx, owner, repo, issueObject)
+	if err != nil {
+		log.Error("failed fetching issue", zap.Error(err))
+		return ctrl.Result{}, err
+	}
+
+	if !issueObject.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("closing issue")
+
+		if gitHubIssue == nil {
+			return ctrl.Result{}, fmt.Errorf("cannot close issue: gitHubIssue is nil")
+		}
+
+		if err := r.CloseIssue(ctx, owner, repo, gitHubIssue); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed closing issue: %v", err)
+		}
+		if err := finalizer.Cleanup(ctx, r.Client, issueObject, r.Log); err != nil {
+			log.Error("failed cleaning up finalizer", zap.Error(err))
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	err = finalizer.Ensure(ctx, r.Client, issueObject, r.Log)
+	if err != nil {
+		log.Error("failed adding finalizer", zap.Error(err))
+		return ctrl.Result{}, err
+	}
+
+	if gitHubIssue == nil {
+		log.Info("creating issue")
+		if err := r.CreateIssue(ctx, owner, repo, issueObject); err != nil {
+			log.Error("failed creating issue", zap.Error(err))
+			return ctrl.Result{}, err
+		}
+
+		gitHubIssue, err = r.FindIssue(ctx, owner, repo, issueObject)
+		if err != nil {
+			log.Error("failed fetching newly created issue", zap.Error(err))
+			return ctrl.Result{}, err
+		}
+		if gitHubIssue == nil {
+			log.Warn("Cannot update status: githubIssue is nil", zap.String("IssueName", issueObject.Name), zap.String("Namespace", issueObject.Namespace))
+		} else {
+			if err := r.UpdateIssueStatus(ctx, issueObject, gitHubIssue); err != nil {
+				log.Error("failed updating issue status", zap.Error(err))
+			}
+		}
+		log.Info("issue created successfully")
+		return ctrl.Result{}, nil
+	} else {
+		log.Info("editing issue")
+		if err := r.EditIssue(ctx, owner, repo, issueObject, *gitHubIssue.Number); err != nil {
+			log.Error("failed editing issue", zap.Error(err))
+			return ctrl.Result{}, err
+		}
+
+		gitHubIssue, err = r.FindIssue(ctx, owner, repo, issueObject)
+		if err != nil {
+			log.Error("failed fetching updated issue", zap.Error(err))
+			return ctrl.Result{}, err
+		}
+
+		if gitHubIssue == nil {
+			log.Warn("Cannot update status: githubIssue is nil", zap.String("IssueName", issueObject.Name), zap.String("Namespace", issueObject.Namespace))
+		} else {
+			if err := r.UpdateIssueStatus(ctx, issueObject, gitHubIssue); err != nil {
+				log.Error("failed updating issue status", zap.Error(err))
+			}
+		}
+		log.Info("issue edited successfully")
+		return ctrl.Result{}, nil
```

Comment:
Same here

Reply:
There also seems to be a lot of duplicate code between the `create` and `update` phases, think about how you can reduce it